benchmark/cublas/blas1/asum.cpp

```diff
@@ -0,0 +1,148 @@
+/***************************************************************************
+ *
+ *  @license
+ *  Copyright (C) Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename asum.cpp
+ *
+ **************************************************************************/
+
+#include "../utils.hpp"
+
+template <typename scalar_t>
+std::string get_name(int size) {
+  std::ostringstream str{};
+  str << "BM_Asum<" << blas_benchmark::utils::get_type_name<scalar_t>() << ">/";
+  str << size;
+  return str.str();
+}
+
+template <typename scalar_t, typename... args_t>
+static inline void cublas_routine(args_t&&... args) {
+  if constexpr (std::is_same_v<scalar_t, float>) {
+    CUBLAS_CHECK(cublasSasum(std::forward<args_t>(args)...));
+  } else if constexpr (std::is_same_v<scalar_t, double>) {
+    CUBLAS_CHECK(cublasDasum(std::forward<args_t>(args)...));
+  }
+  return;
+}
+
+template <typename scalar_t>
+void run(benchmark::State& state, cublasHandle_t* cuda_handle_ptr, index_t size,
+         bool* success) {
+  // init Google-benchmark counters.
+  blas_benchmark::utils::init_level_1_counters<scalar_t>(state, size);
+
+  cublasHandle_t& cuda_handle = *cuda_handle_ptr;
+
+  // Create data
+  std::vector<scalar_t> v1 = blas_benchmark::utils::random_data<scalar_t>(size);
+  // We need to guarantee that cl::sycl::half can hold the sum
+  // of x_v without overflow by making sum(x_v) to be 1.0
+  std::transform(std::begin(v1), std::end(v1), std::begin(v1),
+                 [=](scalar_t x) { return x / v1.size(); });
+
+  scalar_t vr;
+
+  blas_benchmark::utils::CUDAVector<scalar_t> d_x(size, v1.data());
+  blas_benchmark::utils::CUDAScalar<scalar_t> d_r{};
+  CUDA_CHECK(cudaDeviceSynchronize());
+
+  CUBLAS_CHECK(cublasSetPointerMode(cuda_handle, CUBLAS_POINTER_MODE_DEVICE));
+#ifdef BLAS_VERIFY_BENCHMARK
+  // Run a first time with a verification of the results
+  scalar_t vr_ref = reference_blas::asum(size, v1.data(), 1);
+  scalar_t vr_temp = 0;
+  {
+    // Temp result on device copied back upon destruction
+    blas_benchmark::utils::CUDAScalar<scalar_t, true> vr_temp_gpu(vr_temp);
+    CUDA_CHECK(cudaDeviceSynchronize());
```

Comment:
Delete the synchronization here, on line 77. Move synchronization from 91 to 109.